### PR TITLE
Navigation: Use the ListView in the Navigation block inspector controls

### LINKED ISSUE #49986: List View: Open Link UI when inserting a block.
## What problem does this address?
When inserting a block using the OffCanvasEditor component in the Navigation block, a Link UI is opened in the sidebar so that users can select the correct link:
Screenshot 2023-04-21 at 10 11 05

## What is your proposed solution?
We want to retire the OffCanvasEditor component and use the List View in both places, but that would mean modifying the List View to allow the same behaviour as above. There is more discussion on this here: https://github.com/WordPress/gutenberg/issues/48072

### PR DESCRIPTION
**Pull request by scruffian**
## What?
This replaces the OffCanvasEditor with the ListView component in the inspector controls for the Navigation block

## Why?
We shouldn't have two components that do the same thing.

## How?
Implements the new props we have added.

Also adds a wrapping div to allow the list view to scroll horizontally.

## Testing Instructions
1. Open the Site Editor 
2. Select the navigation block
3. Open the inspector controls
4. Confirm that the list view for the navigation block works the same as before

## Screenshots or screencast 
Screenshot 2023-04-17 at 15 37 57

Horizontal scrollbar:
Screenshot 2023-04-17 at 15 38 04

Closes https://github.com/WordPress/gutenberg/issues/49986

### CODE REVIEW

File: packages/block-editor/src/components/list-view/index.js
Comment (scruffian) [inline packages/block-editor/src/components/list-view/index.js]: Maybe if `parentId` is set we should use that to work out what is in `blocks`?

File: packages/block-library/src/navigation/edit/menu-inspector-controls.js
Comment (scruffian) [inline packages/block-library/src/navigation/edit/menu-inspector-controls.js]: We might want to lock this prop.

File: packages/block-editor/src/components/list-view/block.js
Comment (getdave) [inline packages/block-editor/src/components/list-view/block.js]: Why don't you just consume the context in the `BlockSettingsMenu` as well?

I guess it might re-render more often but I don't _think_ it's an expensive component is it?

File: packages/block-editor/src/components/list-view/index.js
Comment (getdave) [inline packages/block-editor/src/components/list-view/index.js]: Should we call this "root" block? Like this is the "roof" that the list view represents?

File: packages/block-editor/src/components/list-view/block.js
Comment (scruffian) [reply packages/block-editor/src/components/list-view/block.js]: Sometimes the BlockSettingsMenu is a LeafMoreMenu, which doesn't live inside the ListView component. I thought it was better to pass the props than to make the LeafMoreMenu a part of the ListView component, so that we can pass in other BlockSettingsMenu components in the future.

File: packages/block-editor/src/components/list-view/index.js
Comment (scruffian) [reply packages/block-editor/src/components/list-view/index.js]: Makes sense to me, but it's called parentId elsewhere in ListView so we'd need to update those.

**Comment by getdave** [comment]
@scruffian Let's try to get https://github.com/WordPress/gutenberg/pull/49433 merged so as to provide more confident about this refactor.

File: packages/block-editor/src/components/list-view/index.js
Comment (talldan) [reply packages/block-editor/src/components/list-view/index.js]: Sounds like a good idea.

The `blocks` prop may need to be deprecated, but it looks like it wouldn't be difficult to do.

File: packages/block-editor/src/components/list-view/index.js
Comment (talldan) [reply packages/block-editor/src/components/list-view/index.js]: I agree `rootClientId` would be good, especially if it ends up being used to determine the blocks in List View as suggested in the other comment.

(I think it's my fault it's called `parentId` 😄 )

File: packages/block-editor/src/components/list-view/block.js
Comment (talldan) [reply packages/block-editor/src/components/list-view/block.js]: It could be a hook (`const [ expand, isExpanded ] = useListViewExpand` (I don't know what to call it)) that uses the context.

I think this is ok though. I think it might be good to look at a hook as part of refactoring the context providers as mentioned in that other PR. Maybe I'll try to tackle that.

File: packages/block-library/src/navigation/edit/menu-inspector-controls.js
Comment (talldan) [reply packages/block-library/src/navigation/edit/menu-inspector-controls.js]: Agree 👍

File: packages/block-library/src/navigation/edit/menu-inspector-controls.js
Comment (talldan) [inline packages/block-library/src/navigation/edit/menu-inspector-controls.js]: Not introduced here, but the boolean props could be shorted to just `showAppender` to save a few bytes.

**Comment by talldan** [comment]
This seems too easy 😄

File: packages/block-editor/src/components/list-view/index.js
Comment (scruffian) [reply packages/block-editor/src/components/list-view/index.js]: I think `parentId` can be a block inside the root though so maybe that's ok. I created a new PR for just the `rootClientId` prop: [List View: Add a rootClientId prop](https://github.com/WordPress/gutenberg/pull/49475#top)

File: packages/block-editor/src/components/list-view/index.js
Comment (scruffian) [reply packages/block-editor/src/components/list-view/index.js]: I'm not sure. In some ways I'd rather keep it as it makes the component more flexible, although we don't use that prop at the moment.

File: packages/block-editor/src/components/list-view/block.js
Comment (scruffian) [reply packages/block-editor/src/components/list-view/block.js]: If we're doing that we can just use the context hook. Updated.

**Comment by scruffian** [comment]
I pulled out the extra prop into another PR to keep the changes as small as possible: https://github.com/WordPress/gutenberg/pull/49475

> This seems too easy 😄

We we did make other changes to ListView first to bring them inline :)

**Comment by scruffian** [comment]
I added a commit to implement the `custom-scrollbars-on-hover` in the inspector. I'd like to get this one in soon if we can so we can start to retire the OffCanvasEditor component.

**Review by ajlende — APPROVED** [review]
This is working well for me!

**Comment by scruffian** [comment]
I think to get the tests to pass we also need to do this: https://github.com/WordPress/gutenberg/pull/49907

**Review by aristath — APPROVED** [review]
Code looks good and works as expected.

> I think to get the tests to pass we also need to do this: https://github.com/WordPress/gutenberg/pull/49907

That PR was just merged, we should rebase this one and see if tests now pass 👍

**Comment by scruffian** [comment]
There's another failing test. More on that here: https://github.com/WordPress/gutenberg/pull/49918

**Comment by scruffian** [comment]
Weird the tests pass locally...

**Comment by andrewserong** [comment]
I was looking at some unrelated tasks and it just occurred to me that with this PR about to land, the id attributes for each list view item should be unique to the list view instance, since there'll be multiple list views on the same page, both referencing the same blocks? I've opened up a small code quality PR in https://github.com/WordPress/gutenberg/pull/49944 to explore making those ids unique.

**Comment by scruffian** [comment]
We also need to address https://github.com/WordPress/gutenberg/issues/49986 before we can merge this.

**Comment by jasmussen** [comment]
~Will this remove the drilldown behavior? I'm not sure the inspector is the best place for showing nesting like this. I'd love @SaxonF be looped into these changes, as they seem to be quite a departure from initial concepts.~

Edit: Actually, ignore me. I had not had my morning coffee yet.

File: packages/block-editor/src/components/list-view/link-ui.js
Comment (jeryj) [inline packages/block-editor/src/components/list-view/link-ui.js]: Should this file be hooked up via the navigation block instead? I don't have enough context, but it feels like if this is only for the navigation that it should be within the navigation instead.

File: packages/block-editor/src/components/list-view/block-contents.js
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/block-contents.js]: Similar feedback to this comment: https://github.com/WordPress/gutenberg/pull/49417#discussion_r1189115306 — is this change baking in too much knowledge about particular blocks into the `ListView` component?

Would it be possible to split out the link UI behaviour so that it's added to the list view by the consumer of the component, a bit like how there's now the `renderAdditionalBlockUI` prop as of https://github.com/WordPress/gutenberg/pull/50465?

File: packages/block-editor/src/components/list-view/link-ui.js
Comment (andrewserong) [reply packages/block-editor/src/components/list-view/link-ui.js]: I agree, I think it could be preferable to keep this sort of logic in the block, to try to avoid adding too much extra complexity to the `ListView` component itself, if possible. I've added a similar comment in https://github.com/WordPress/gutenberg/pull/49417#discussion_r1189275377.

Overall, I think it'd be good if we can avoid the list view having to know too much about particular blocks.

File: packages/block-library/src/navigation/editor.scss
Comment (MaggieCabrera) [inline packages/block-library/src/navigation/editor.scss:664]: I changed the parameters because it was doing the opposite (hiding on hover instead of showing)

File: packages/block-editor/src/components/list-view/appender.js
Comment (scruffian) [inline packages/block-editor/src/components/list-view/appender.js]: We should remove this since we're going to do it this way instead: https://github.com/WordPress/gutenberg/pull/50510

File: packages/block-library/src/navigation/edit/menu-inspector-controls.js
Comment (scruffian) [inline packages/block-library/src/navigation/edit/menu-inspector-controls.js]: I think this should be a separate function definition

File: packages/block-library/src/navigation/edit/menu-inspector-controls.js
Comment (scruffian) [inline packages/block-library/src/navigation/edit/menu-inspector-controls.js:135]: Can we avoid this extra div?

Comment (MaggieCabrera) [reply packages/block-library/src/navigation/edit/menu-inspector-controls.js:135]: no, we need the classname on a div, not the table element for the horizontal scroll to work

**Comment by MaggieCabrera** [comment]
The e2e failing tests here are because we need https://github.com/WordPress/gutenberg/pull/50510 for the inserter order issue. Should we wait for that to land or skip those tests and re-enable them on #50510?

**Review by draganescu — DISMISSED** [review]
This tests well for me and the code does the job of making the leafmoremenu a component that is composed from the various places that use ListView.

The leaf more menu in the navigation screen of the site editor does not perfectly work. Particularly the add submenu:

- does not expand the link with the added submenu item
- the added submenu item shows up LinkUI - which is as broken as always in this context

I will approve this anyway as I think it's a good step forward.

File: packages/block-editor/src/components/list-view/private-list-view.js
Comment (draganescu) [inline packages/block-editor/src/components/list-view/private-list-view.js]: Are we sure we want to remove this? @talldan doesn't this remain a viable alternative, giving ListView a list of blocks? I think we're corneting ListView intro a place where it can only work along with an instanced block editor tree. I am suspicious if this actually improves ListView or takes away.

I'd be happy in the future to have some need to pass a tree of blocks and for it to work.

File: packages/block-editor/src/components/list-view/index.js
Comment (ajlende) [inline packages/block-editor/src/components/list-view/index.js:275]: I removed this in my commit because `BlockSettingsDropdown` is the default value, and it meant that there were fewer things that needed to be imported here.

This change does mean that it can be overridden by props passed to `ListView` now, so if that shouldn't be possible, then it needs to be added back as `undefined` or something.

Comment (scruffian) [reply packages/block-editor/src/components/list-view/index.js:275]: Yeah we can't do that because the point it to restrict this from being prop being used by third parties. `undefined` should work though, I'll push that now.

Comment (scruffian) [reply packages/block-editor/src/components/list-view/index.js:275]: Done in 74a8ac27037f87b7daf001213565aee627e1e731

File: packages/block-editor/src/components/list-view/index.js
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/index.js]: Why is `unlock` being used here? `PrivateListView` is in the same package, so can't it be imported directly without having to use the `unlock` API?

File: packages/block-editor/src/components/list-view/private-list-view.js
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/private-list-view.js]: Apologies if this has already been discussed somewhere, but I'm wondering why this component is being moved to another file? It creates quite a big diff, which means that we could lose some useful history in the `ListViewComponent`. Is it possible to keep `ListViewComponent` in `index.js`?

Comment (jeryj) [reply packages/block-editor/src/components/list-view/private-list-view.js]: We didn't want to move it, but webpack was causing a build error with having both exports in the `index.js` file. It was working fine, then all of a sudden was not working on anyone's machine. It seemed to be a build issue and not related to the code itself, and moving the file fixed it. At least, that's my understanding from @ajlende. He may be able to provide more context.

File: packages/block-editor/src/components/list-view/private-list-view.js
Comment (talldan) [reply packages/block-editor/src/components/list-view/private-list-view.js]: I'm not sure when `blocks` was introduced, but it is a little open to issues:
- Some selectors that return block trees only contain uncontrolled blocks
- ListView is optimized to work with a `clientIdsTree` rather than a full block tree, passing in a full block tree is overly expensive.

> I think we're corneting ListView intro a place where it can only work along with an instanced block editor tree.

It's a bit of an existential issue. I'm not sure what use cases there would be for it work with a non-instanced block editor.

File: packages/block-editor/src/components/list-view/private-list-view.js
Comment (scruffian) [reply packages/block-editor/src/components/list-view/private-list-view.js]: I moved it back in [3fe3bf9](https://github.com/WordPress/gutenberg/pull/49417/commits/3fe3bf99b7507df47a2c24f3aee04ec2322f94c2)

File: packages/block-editor/src/components/list-view/index.js
Comment (scruffian) [reply packages/block-editor/src/components/list-view/index.js]: Fixed in [3fe3bf9](https://github.com/WordPress/gutenberg/pull/49417/commits/3fe3bf99b7507df47a2c24f3aee04ec2322f94c2)

**Comment by andrewserong** [comment]
Nice work landing this one, folks! 🎉

**Pull request merged**
